Support for USB controller on STM32H7A3

aurel32 opened the pull request:
This patchset adds support for the USB controller on the STM32H7A3 SoC, which differs from the already supported STM32H7 SoC by the lack of the second USB controller.

aurel32 committed:
drivers: usb: device: stm32: do not manually enable the clock on STM32H7

Do not manually enable the USB clock on STM32H7, as it is already done
in usb_dc_stm32_clock_enable() using the values in the device tree.

This partially fixes the build for STM32H7 devices with a single USB
controller.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
drivers: usb: device: stm32: fix disabling the ULPI clock on STM32H7X

The ULPI clock uses slightly different names on STM32H7X as those
SoC can have more than one USB controller.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
drivers: usb: device: stm32: fix a wrong comment STM32H7X and ULPI clock

On STM32H7X SoCs, both USB controllers are HS capable, however it is not
possible to connect an ULPI PHY to the USB2 one, which limit it to FS.
Therefore the ULPI clock in sleep mode has to be disabled for USB2 in
all cases. The ULPI clock for USB1 is already disabled when needed in
the usb_dc_stm32_clock_enable() function like for other SoCs.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
dts: arm: st: h7: add USB OTG HS controller

Add the USB OTG HS controller to the device tree and the corresponding
FS PHY.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
boards: arm: nucleo_h7a3zi_q: add USB support

The Nucleo H7A3ZI-Q board has a micro USB connector connected to the USB
OTG HS controller. Enable it and update the documentation.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -305,6 +305,10 @@ static int usb_dc_stm32_clock_enable(void)
 #if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_usbphyc)
 	LL_AHB1_GRP1_EnableClock(LL_AHB1_GRP1_PERIPH_OTGHSULPI);
 	LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_OTGPHYC);
+#elif defined(CONFIG_SOC_SERIES_STM32H7X)
+	/* Disable ULPI interface (for external high-speed PHY) clock */
+	LL_AHB1_GRP1_DisableClock(LL_AHB1_GRP1_PERIPH_USB1OTGHSULPI);
+	LL_AHB1_GRP1_DisableClockSleep(LL_AHB1_GRP1_PERIPH_USB1OTGHSULPI);
 #else
 	/* Disable ULPI interface (for external high-speed PHY) clock */
 	LL_AHB1_GRP1_DisableClock(LL_AHB1_GRP1_PERIPH_OTGHSULPI);
```
For my information, is the disable step required ?

erwango replied:
Thanks for the heads up.
Can we then keep only the disable in sleep mode ?

aurel32 replied:
I have just pushed a new version with a commit removing the corresponding lines.

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -396,12 +396,15 @@ static int usb_dc_stm32_init(void)
 #if defined(CONFIG_SOC_SERIES_STM32H7X)
-	/* Currently assuming FS mode. Need to disable the ULPI clock on USB2.
-	 * Need to make this dependent on HS or FS config.
+#if defined LL_AHB1_GRP1_PERIPH_USB2OTGHSULPI
+	/* The USB2 controller only works in FS mode, but we still need to
+	 * disable its ULPI clock. As the USB1 controller is an HS one, this is
+	 * done in the common path.
```
Can we somehow use device tree to get this information ?

aurel32 replied:
> Can we somehow use device tree to get this information ?

It's not really clear to me. The USB controller is declared as a `st,stm32-otgfs` in the device tree because it can't support HS physically, but from the register point of view it is actually a HS controller with an ULPI interface.

aurel32 replied:
I realized that this patch mixes unrelated comment fixes and code changes. I have therefore split this one in two parts:
- One that fixes the existing comment about the second USB controller, but doesn't change any code. The part _Need to make this dependent on HS or FS config_ is wrong as the second controller can only work in FS mode with the internal FS PHY.
- One that limit the changes on the second USB controller to the SoCs that have two USB controllers.

aurel32 replied:
The problem there is to make the link between the USB controller and the USB PHY. That something that can be modeled in the device tree, however what we want here is to **disable** the **not used** PHY clock, while the device tree models the devices that are used.

After thinking a bit more I have found a way to use the device tree, by checking for the `st-stm32-otgfs` compat property of the USB controller. That said it hardcodes in a bit more places that the USB1 controller is HS and the USB2 controller is FS on STM32H7x SoC. That will break if ST decides to create a STM32H7x SoC with 2 HS controllers.

aurel32 committed:
drivers: usb: device: stm32: only disable the ULPI clock for USB FS

On SoCs with two USB controller, Zephyr selects the FS one, and in that
case the ULPI clock needs to be disabled in sleep mode for the
controller to work.

On SoCs with a single USB HS controller, this operation is not needed.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -396,8 +396,9 @@ static int usb_dc_stm32_init(void)
 #if defined(CONFIG_SOC_SERIES_STM32H7X)
-	/* Currently assuming FS mode. Need to disable the ULPI clock on USB2.
-	 * Need to make this dependent on HS or FS config.
+	/* The USB2 controller only works in FS mode, but we still need to
+	 * disable its ULPI clock. As the USB1 controller is an HS one, this is
+	 * done in the common path.
```
I'm still not getting why this step is required as this clock is off by default.
Is there a piece of code that activates it indirectly before this point?

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -396,13 +396,15 @@ static int usb_dc_stm32_init(void)
 #endif /* CONFIG_USB_DEVICE_SOF */
 
 #if defined(CONFIG_SOC_SERIES_STM32H7X)
+#if defined LL_AHB1_GRP1_PERIPH_USB2OTGHSULPI
```
Anyway to use device tree bits to infer this ?

aurel32 committed:
drivers: usb: device: stm32: do not disable ULPI clock in run mode

The ULPI clock is disabled by default in run mode, there is no need to
disable it again. It is however enabled by default in sleep/low power
mode, so it needs to be disabled.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

erwango reviewed (approved):
LGTM, thanks!
